### DISCUSSION TITLE
Supabase requests are being hit even tho I am caching as much as I can using NextJS?

### DISCUSSION THREAD

**Discussion by Tigatok** [supabase/supabase#28157] [Questions] [answered]
I am trying to figure out if there is a bug in my code with how I am handling caching, or if I am just missing a concept here. So I have a NextJS application running in Vercel. Its using the app router. I have a page that is a server component that calls the following function:

```javascript
const getLocationRatings = async (): Promise<MyDataType[]> => {
  const supabase = createClient();

  const { data, error }: PostgrestResponse<MyDataType> =
    await supabase.rpc("get_my_data");
  return data;
};
```

I've updated my supabase server createClient to use a fetch client that should use NextJS caching:

```javascript
export const createFetch =
  (options: Pick<RequestInit, "next" | "cache">) =>
  (url: RequestInfo | URL, init?: RequestInit) => {
    return fetch(url, {
      ...init,
      ...options,
    });
  };

export function createClient() {
  const cookieStore = cookies();

  return createServerClient(
    process.env.SUPABASE_URL!,
    process.env.SUPABASE_SERVICE_KEY!,
    {
      global: {
        fetch: createFetch({
          next: {
            revalidate: 300,
            tags: ["supabase"],
          },
        }),
      },
      cookies: {
        get(name: string) {
          return cookieStore.get(name)?.value;
        },
        set(name: string, value: string, options: CookieOptions) {
          try {
            /* eslint-disable-next-line */
            cookieStore.set({ name, value, ...options });
          } catch (error) {
            // The `set` method was called from a Server Component.
            // This can be ignored if you have middleware refreshing
            // user sessions.
          }
        },
        remove(name: string, options: CookieOptions) {
          try {
            /* eslint-disable-next-line */
            cookieStore.set({ name, value: "", ...options });
          } catch (error) {
            // The `delete` method was called from a Server Component.
            // This can be ignored if you have middleware refreshing
            // user sessions.
          }
        },
      },
    }
  );
}
```

I then have 2 server actions, that when executed, invalidate the cache tag "supabase". This works, however, on every page load, I am still seeing my requests to Supabase increase. I also notice that my Storage in Supabase where it shows cached items, appears to say nothing is cached.

Am I a bit confused with something? My goal is to always return the same data, unless I revalidate it (which happens on a webhook firing from Supabase).

<img width="1161" alt="Screenshot 2024-07-23 at 10 01 59 PM" src="https://github.com/user-attachments/assets/39d43048-443c-49da-b246-cdf65b79b279">

**Comment by ahofmeister** [comment]
Coming here for the very same question. 

Having 3 different components: 

```
export default async function Index() {

    return (
        <div>
            <SupabaseFetchWithClient />
            <SupabaseFetchWithServer />
           <NormalFetch />
        </div>
    );
}
```

The client fetching component looks like this: 

```
import {createClient} from "@/utils/supabase/client";

export default async function SupabaseFetchWithClient() {

    const supabase = createClient();
    const {data} = await supabase.from("transactions").select().limit(2)

    return (
        <div>
            <pre>{JSON.stringify(data, null, 2)}</pre>
        </div>

    );
}

```


The server fetching component looks like this

```
import {createClient} from "@/utils/supabase/server";

export default async function SupabaseFetchWithServer() {

    const supabase = createClient();
    const {data} = await supabase.from("transactions").select().limit(2)

    return (
        <div>
            <pre>{JSON.stringify(data, null, 2)}</pre>
        </div>
    );
}

```


The first and third one works as intended by hitting the cache immediately. But with the server component it always prints out 



```
 GET / 200 in 125ms
 │ GET https://<project_id>.supabase.co/rest/v1/transactions?select=*&limit=2 200 in 65ms (cache skip)
 │ │ Cache skipped reason: (auto no cache)
 │ GET https://api.vercel.app/products/1 200 in 85ms (cache skip)
 │ │ Cache skipped reason: (auto cache)
 ```
 
 To setup everything I started with `npx create-next-app -e with-supabase`
 
 I followed https://supabase.com/blog/fetching-and-caching-supabase-data-in-next-js-server-components but it seems a bit outdated.

**Reply by Tigatok** [reply]
So my problem was solved by using the right package, lol. I was using supabase/ssr which uses cookies, and that opts out of nextjs caching. So I realised the package they use in the blog article is actually supabase-js and that one doesn’t use cookies. 

Hope that helps.

**Reply by ahofmeister** [reply]
I see. However all the (new) starters and templates use the ssr package. As I said, the blog article seems a bit outdated. Do you have any docs that state, that using cookies is a opt out for caching?

**Reply by ahofmeister** [reply]
nevermind, found them: https://supabase.com/docs/guides/auth/server-side/nextjs?queryGroups=router&router=app

>Note that cookies is called before any calls to Supabase, which opts fetch calls out of Next.js's caching. This is important for authenticated data fetches, to ensure that users get access only to their own data.

See the Next.js docs to learn more about [opting out of data caching](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#opting-out-of-data-caching).

Is using supabase-js the only solution?

**Reply by ahofmeister** [reply]
Okay, I got it working with `supabase-js`. However, every time the server restarts, I am logged out. :/ Not perfect, but the caching works.  This also is true for every new route I visit after a server restarts....

Okay, after reading into it. Makes sense. The server cannot access the session, as the session is stored in LocalStorage.

**Reply by Tigatok** [reply]
@ahofmeister I would be pretty careful about using it in that capacity, as I believe you want to use the supabase-ssr package if you are using any of the authentication. The reason I am not worried about making calls without cookies is because I am just doing server-to-server calls, which use a secret key and can be fully cached.

**Reply by Tigatok** [reply]
I am not sure if I fully understand/agree why they called the new package supabase-ssr. In my opinion, that was the main confusing point here. As what I am doing, using `supabase-js` I am still doing SSR.

**Reply by ahofmeister** [reply]
Well, I am pretty confused about what is the right thing to do here. :/ Appreciate your feedback!

**Comment by Tigatok** [comment, marked as answer]
So my problem was solved by using the right package, lol. I was using supabase/ssr which uses cookies, and that opts out of nextjs caching. So I realised the package they use in the blog article is actually supabase-js and that one doesn’t use cookies. 

Hope that helps
